Feature/branch name in reftable repos

nikolaosginos opened the pull request:
<!-- Please read CONTRIBUTING.md before submitting a pull request -->

Fixes #12491

## Proposed changes

In git reftable repos , the HEAD file names the current branch ".invalid", since that information is stored in the binary reftable format instead. This causes GitExtenstions `GetSelectedBranchFast` to return the branch name ".invalid".

- When the branch name is ".invalid", treat it as if the function failed to retrieve the current branch name
- When the "fast" function fails to retrieve the current branch name, use the traditional method instead

## Screenshots <!-- Include variants with higher scaling, e.g. 150% or 200%. Remove this section if PR does not change UI -->

### Before

<img width="655" height="278" alt="image" src="https://github.com/user-attachments/assets/02b79a76-1e07-4407-a2b9-dffdda0f9771" />

### After

<img width="651" height="253" alt="image" src="https://github.com/user-attachments/assets/85846db5-b4ae-4b6d-9f62-957c918febb1" />


## Test methodology <!-- How did you ensure quality? -->

- Manual tests with the instructions provided in the issue
- Used a build of GitExtensions that contains these changes productively for a week

## Test environment(s) <!-- Remove any that don't apply -->

- GIT 2.51
- Windows 11 24H2

<!-- Mention language, UI scaling, or anything else that might be relevant -->

## Merge strategy

<!-- Change the following if the merge strategy should be changed:
- Squash merge (maintainer to decide merge message, PR submitter should cleanup commits/messages at PR approval).
- Rebase merge (PR submitter must change the commit message for the last commit).
- Merge commit. (PR submitter to rebase and squash before merges).
- To be decided later.
The maintainer may still request the contributor to squash and rebase, to make sure that merges and commit messages are clarified.
-->

I agree that the maintainer squash merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

nikolaosginos commented on src/app/GitCommands/Git/GitModule.cs:
```diff
@@ -2887,7 +2887,8 @@ public static string GetSelectedBranchFast(string? repositoryPath, bool emptyIfD
 
-            if (!headFileContents.StartsWith(prefix))
+            // the head file contains ".invalid" if the repo uses the reftable backend
+            if (!headFileContents.StartsWith(prefix) || headFileContents.StartsWith($"{prefix}.invalid"))
```
`git rev-parse --show-ref-format` does that. It returns "files" in classic repos and "reftable" in reftable repos. I can use that instead. I suppose this is the indented way to detect it. Can you give me a hint on what the preferred way to handle that information is in the GitExtensions project? Is there some class that contains basic information about the currently opened repo where I can store it when the repo gets opened (`GitModule`?)? Or should I execute that command on demand here?

gerhardol replied:
`GetSelectedBranchFast` is intended to be used without any Git calls, to not delay the UI.
All these commands adds up and suddenly it takes many seconds extra to open a repo
GetSelectedBranchFast is used outside of GitModule in some situations where the branch is not so important

Git commands can be cached, but my question was if there is a way to get the reftable info from the filesystem?

nikolaosginos commented on src/app/GitCommands/Git/GitModule.cs:
```diff
@@ -2911,8 +2912,10 @@ public string GetSelectedBranch(bool emptyIfDetached = false)
+
             return result.ExitedSuccessfully
-                ? result.StandardOutput
+                ? result.StandardOutput[prefix.Length..].TrimEnd()
```
I was unsure about that change at first but I really think it shouldn't be the full ref. This `GetSelectedBranch` method calls `GetSelectedBranchFast` first. If that is successful, it exits early returning only the branch name, not the full ref. I would be inconsistent to return the full ref here.

I looked at the references of `GetSelectedBranch` and all usages seem to treat the result as the simple branch name. I also found one usage where `GitRefName.GetFullBranchName` was explicitly called on its result to get the full ref.

When I set a breakpoint there and worked with non-reftable repos, it never even reached that breakpoint because the "fast" method always worked. So I suspect that this line was indeed wrong but it didn't cause problems because the function always exited early with the fast method.

Please correct me if I am wrong. This is my first contribution here. But it looks correct to me. If I revert that line. The result is the following:

<img width="709" height="270" alt="image" src="https://github.com/user-attachments/assets/31e62d09-1701-4e8b-85ae-5728d4fe6817" />

The branch name in the UI includes "refs/heads/" and the revision graph does not detect the branch actually being checked out (green branch marker should be bold).

RussKie commented on src/app/GitUI/RepositoryCurrentBranchNameProvider.cs:
```diff
@@ -26,6 +26,12 @@ public string GetCurrentBranchName(string repositoryPath)
+
+        if (string.IsNullOrWhiteSpace(branchName))
+        {
+            branchName = new GitModule(repositoryPath).GetSelectedBranch();
```
`GetSelectedBranch` starts with invoking `GetSelectedBranchFast`:
https://github.com/gitextensions/gitextensions/blob/9dd29b2edefafc9c6a46861895776696dae0b1eb/src/app/GitCommands/Git/GitModule.cs#L2900

I don't think this is the most optimal way of addressing the issue. Instead, we should pass `Func<IGitModule>` into the .ctor (the same way we do in many other places), and then call
```cs
string branchName = getModule().GetSelectedBranch();
```

`GetSelectedBranchFast` should be turned into a private (or internal for testing purposes) method.

gerhardol replied:
> GetSelectedBranchFast should be turned into a private (or internal for testing purposes) method.

GetSelectedBranchFast is used in other modules when the branch name is not critical (for display) and speed is important, so it should remain public.

gerhardol commented on src/app/GitUI/CommandsDialogs/BrowseDialog/DashboardControl/UserRepositoriesListController.cs:
```diff
@@ -47,12 +50,16 @@ public void ClearCache()
             }
 
-            return GitModule.GetSelectedBranchFast(path);
+            string branchName = module.GetSelectedBranch();
```
With this change, a Git command could be required for many repos, like submodules where no branch is checked out.
I have 150 modules in my dashboard and 100 of them are submodules, so GE will require 10s extra or so.

mstv commented:
I have not dealt with the details.
Though I guess that a repo either uses reftable or the classic format.
So, it should suffice to detect it once and then always run the adequate git command without the need for a second as fallback.

gerhardol commented:
> So since we don't want to call a git command for performance reasons, this approach for detecting it should be fine.

.invalid is not possible as a Git name for a branch, so the detection is probably OK.
I guess Fast could return an empty string if the name is .invalid.

For the "full" usage with many queries I guess there could be a setting to always query (even if GE has too many settings already).
My case is not the most common....

nikolaosginos commented:
I have added comments and tests as requested. Are the tests okay? There were no existing tests for `GetSelectedBranch`. My changes did not break any of the other existing tests. Just one unrelated test fails before and after my changes:(`DiffHighlightServiceTests.MarkInlineGap`).

> I guess Fast could return an empty string if the name is .invalid.

I did not do this because the caller would not know if the command returned an empty string because the repo is in a detached state or because it is not able to retrieve the branch name because the repo uses reftables. Those cases have to be treated differently by the caller.

But if we implemented some other way of checking for the reftable backend (e. g. reading `.git/config` or checking for the `.git/reftable` folder) we would be able to check first and not even try the fast function. I don't have a strong opinion on this. Checking .config is a little more complex since it involves parsing the file. Checking for the 'reftable' subfolder seems like a valid option though. I also tried migrating back and forth and it gets deleted when switching back to the regular files backend. So this might be a safe way to check as well. And we would not have to rely on trying the fast method and interpreting its ".invalid" return value. I think checking by looking for the subfolder before calling one of the `GetSelectedBranch` functions would be easier to read and maintain and it feels a little less hacky. What do you think?

> For the "full" usage with many queries I guess there could be a setting to always query (even if GE has too many settings already). My case is not the most common....

If this would be accepted, I will gladly add it. I don't like the dashboard showing ".invalid" for all repos, but as mentioned earlier I could not find a reliable way to get it without calling git.

This pull request in its current state would fix the issue that you can see in the screenshots. I think of this issue as a deal breaker for reftable repos. Would you prefer it if I keep this pull request like this for now to get the fix merged? I could then create another issue and pull request for the dashboard and the "recently used" menus by falling back to the git command. Or should I keep doing those changes in this pull request?

gerhardol commented:
>  I don't like the dashboard showing ".invalid" for all repos, but as mentioned earlier I could not find a reliable way to get it without calling git.

You could just ignoring returning `.invalid` to not display anything at all.

But I am OK with a setting if it is default off.
I guess you have basically written the code already in the previous commits...

If a repo supports reftable could be read with git-config or checking .git/reftable

If you use reftable I assume you work with large repos and will see other performance issues with GE...

mstv commented on src/app/GitCommands/Git/GitModule.cs:
```diff
@@ -2912,7 +2912,7 @@ public string GetSelectedBranch(bool emptyIfDetached = false)
 
             return result.ExitedSuccessfully
-                ? result.StandardOutput
+                ? result.StandardOutput["refs/heads/".Length..].TrimEnd()
```
Does this mean, the result has always been wrong? ... and it has not been noticed because we have never reached this execution branch?

Consider using the pseudo-const `GitRefName.RefsHeadsPrefix`.

mstv commented on tests/app/UnitTests/GitCommands.Tests/GitModuleTests.cs:
```diff
@@ -585,6 +585,28 @@ public void BatchUnstageFiles_should_work_as_expected(GitItemStatus[] files, str
+        [TestCase(true)]
+        public void GetSelectedBranch_should_return_requested_result_in_detached_state(bool emptyIfDetached)
+        {
+            _executable.StageOutput("symbolic-ref --quiet HEAD", "", 1);
```
```suggestion
            _executable.StageOutput("symbolic-ref --quiet HEAD", "", exitCode: 1);
```

gerhardol commented on src/app/GitCommands/Git/GitModule.cs:
```diff
@@ -2850,7 +2850,8 @@ public void UnstageFileToRemove(string file)
+        /// <remarks>Dirty but fast. This sometimes fails. In reftable repos, it always returns ".invalid".</remarks>
         public static string GetSelectedBranchFast(string? repositoryPath, bool emptyIfDetached = false)
         {
             if (string.IsNullOrEmpty(repositoryPath))
```
Maybe in a followup
reftable is set in the config:
https://git-scm.com/docs/reftable#_repository_format

So in that case the file should never be read at all.

gerhardol reviewed (commented):
I want to make it configurable to read branchnames in the menu (the fast approach is also too slow forme).
Maybe add a separate setting for the menu?
(Dashboard is OK for me.)
This could be off by default for reftable.

gerhardol commented:
> A setting for Dashboard and Menu already exists. Do you want to split it into two settings?

I would prefer to have a shorter list (like 30) in the menu and ~300 i Dashboard, but as the list is saved back, it is trickier.
So a split setting is OK for me.

> I would like to be able to see the branch names for reftable repos as well. But this would require falling back to the regular method if the fast method returns ".invalid" (as it does here for now). If there are no reftable repos this would not have a noticeable effect. Would you still want to have a setting to turn it on or off specifically for reftable repos?

The setting may be static and default on but be deactivated manually with many repos or reftable. (I have hardcoded that myself.) I prefer to not add more settings, it is hard to find them as it is.

But the reftable extension can be read from settings (changes outside GE can be ignored, reread at refresh) (the value is cached per session, reread when opening a repo). 
Use this to skip Fast and always do the full Git processing.

If the settings for reftable is removed when default, it has to be handled then, just as the shortcut reading files is plugged here.

RussKie commented on src/app/GitCommands/Git/GitModule.cs:
```diff
@@ -2899,7 +2900,7 @@ public string GetSelectedBranch(bool emptyIfDetached = false)
             string head = GetSelectedBranchFast(WorkingDir, emptyIfDetached);
 
-            if (!string.IsNullOrEmpty(head))
+            if (!string.IsNullOrEmpty(head) && !head.Equals(".invalid"))
```
Oh, that's a good catch. @nikolaosginos a more idiomatic .NET way is to write as `if (!string.IsNullOrEmpty(head) && head != ".invalid")` since we want a case-sensitive comparison.

mstv commented:
Could you resolve the MC, please?
The following steps should work :
1. remove the namespace blocks in the three files
2. squash
3. rebase on master
